src/lib/TopLevel.hs

```diff
@@ -169,7 +169,10 @@ initializeBackend backend = case backend of
   _ -> error "Not implemented"
 
 arrayVars :: HasVars a => a -> [Var]
-arrayVars x = map (\(v@(Name ArrayName _ _), (ty, _)) -> v :> ty) $ envPairs (freeVars x)
+arrayVars x = foldMap go $ envPairs (freeVars x)
+  where go :: (Name, (Type, BinderInfo)) -> [Var]
+        go (v@(Name ArrayName _ _), (ty, _)) = [v :> ty]
+        go _ = []
```

Comment:
Um, that looks incorrect to me? We should not have any free non-array variables here!

Reply:
Hm, really? Maybe there's a better fix then. In practice it seems like every global is treated as free, so if the expression has any global symbols in it (like `Ord`, or `(&)`) it traps on a pattern match failure.

Reply:
Should global variables like that not be treated as free? Or is it ok for `GlobalName`s to show up here, and any other name cause an error? (Or is it bad that global names are showing up here?)

Reply:
Hmmm I mean I think you might be right in that we might start seeing global non-array vars due to newtype binders, but in general we should still check that we don't have anything else, because simplification should have inlined them away!